_pytest/config.py

```diff
@@ -1257,9 +1257,13 @@ def get_common_ancestor(args):
 
 
 def get_dirs_from_args(args):
-    return [d for d in (py.path.local(x) for x in args
-                        if not str(x).startswith("-"))
-            if d.exists()]
+    possible_dirs = [
+        py.path.local(str(arg).split('::')[0])
+        for arg in args
+        if not str(arg).startswith('-')
+    ]
+
+    return [d for d in possible_dirs if d.exists()]
```

Comment:
IMHO what would improve readability would be to use small functions that act as documentation to explain all that's being handled there:

```python
def get_dirs_from_args(args):
    def is_option(x):
        return str(x).startswith('-')

    def get_file_part_from_node_id(x):
        return str(arg).split('::')[0]

    possible_dirs = (
        py.path.local(get_file_part_from_node_id(arg))
        for arg in args
        if not is_option(arg)      
    )
    return [d for d in possible_dirs if d.exists()]
```

Also I just noticed that although the function is named `get_dirs_from_args`, it actually returns directories and filenames. 😁